File: docs/examples/python/provenance.py
"""Run the Python provenance, staleness, and dependency listings.

Each ``[name]`` ... ``[/name]`` region is included verbatim by the
documentation build, so keep regions self-contained and readable. Usage::

    POSTPROJECT_LIBRARY=/path/to/libpostproject.so python provenance.py WORK_DIRECTORY

The work directory is prepared by ``prepare-workdir.cmake``.
"""

from __future__ import annotations

import hashlib
import sys
from pathlib import Path

from postproject import (
    Activity,
    ActivityEdge,
    ActivityId,
    ActivitySpec,
    AgentIdentity,
    ArtifactEvaluation,
    ArtifactKnowledgeState,
    ArtifactReasonKind,
    ArtifactReproducibility,
    ArtifactReproducibilityIssueKind,
    AssetId,
    Dependency,
    DependencyMatch,
    DependencySet,
    DependencySetStatus,
    ExternalIdentifier,
    Fingerprint,
    Production,
    Representation,
    RepresentationId,
    RepresentationKind,
    ToolIdentity,
)


# [activity-snapshots]
def record_proxy_generation(
    production: Production,
    original_id: RepresentationId,
    proxy_id: RepresentationId,
) -> ActivityId:
    with production.transaction() as transaction:
        activity_id = transaction.create_activity(
            ActivitySpec(
                "org.postproject:transcode",
                inputs=(ActivityEdge(original_id, "org.postproject:primary"),),
                outputs=(ActivityEdge(proxy_id, "org.postproject:proxy"),),
                tool=ToolIdentity("Example Transcoder", "3.2"),
                agent=AgentIdentity(
                    "Render node 4", ExternalIdentifier("com.example.host", "node-4")
                ),
            )
        )
    return activity_id


def print_fingerprint_snapshots(edge: ActivityEdge) -> None:
    # Commit captured the fingerprints each edge representation had then.
    if edge.snapshot is None:
        return
    for fingerprint in edge.snapshot.fingerprints:
        print(
            f"  {edge.representation_id} {fingerprint.algorithm}: "
            f"{fingerprint.value.hex()[:16]}..."
        )


def print_activities(production: Production) -> tuple[Activity, ...]:
    activities = production.activities
    for activity in activities:
        tool = activity.tool.name if activity.tool else "-"
        agent = activity.agent.name if activity.agent else "-"
        print(f"{activity.kind} by {tool} on {agent}")
        for edge in activity.inputs + activity.outputs:
            print(f" {edge.role or '-'}")
            print_fingerprint_snapshots(edge)
    return activities


def print_lineage(production: Production, original_id: RepresentationId) -> None:
    for activity in production.activities_consuming[original_id]:
        print(f"consumed by {activity.id}")
    for descendant_id in production.provenance_descendants[original_id]:
        print(f"descendant {descendant_id}")

    cursor = None
    while True:
        page = production.provenance_descendants_page(
            original_id, max_depth=8, max_representations=1000, limit=100, cursor=cursor
        )
        for match in page.items:
            print(f"descendant {match.representation_id} at depth {match.depth}")
        cursor = page.next_cursor
        if cursor is None:
            break


# [/activity-snapshots]


# [stale-after-change]
def record_fingerprints(
    production: Production, representation: Representation, path: Path
) -> None:
    # Observations in the caller's own fingerprint domain.
    digest = hashlib.sha256(path.read_bytes()).digest()
    (resource,) = representation.resources
    with production.transaction() as transaction:
        transaction.record_resource_fingerprint(
            resource.id, Fingerprint("example-sha256", 1, digest)
        )
        transaction.record_representation_fingerprint(
            representation.id,
            Fingerprint(
                "example-sha256-representation", 1, hashlib.sha256(digest).digest()
            ),
        )


def explain_proxy(
    production: Production, proxy_id: RepresentationId
) -> tuple[ArtifactEvaluation, ArtifactReproducibility]:
    evaluation = production.evaluate_artifact(
        proxy_id, max_depth=64, max_representations=1000
    )
    print(f"proxy is {evaluation.state.name}")
    for reason in evaluation.reasons:
        print(f"  {reason.kind.name} on {reason.representation_id}")
        if reason.kind is ArtifactReasonKind.FINGERPRINT_CHANGED:
            assert reason.snapshot_value is not None
            assert reason.current_value is not None
            print(
                f"  {reason.fingerprint_algorithm}: "
                f"{reason.snapshot_value.hex()[:16]} -> {reason.current_value.hex()[:16]}"
            )

    # What is still missing to regenerate the proxy exactly.
    reproducibility = production.artifact_reproducibility(proxy_id)
    for issue in reproducibility.issues:
        print(f"  not reproducible: {issue.kind.name}")
    return evaluation, reproducibility


# [/stale-after-change]


# [dependency-set]
def record_edit_dependencies(
    production: Production,
    edit_id: RepresentationId,
    camera_asset_id: AssetId,
    original_id: RepresentationId,
    proxy_id: RepresentationId,
) -> DependencySet | None:
    # Always record the complete set: it replaces the previous observation.
    dependencies = (
        Dependency(
            "org.example:clip-reference",
            camera_asset_id,
            "rushes/A001.mov",
            resolved_representation_id=original_id,
        ),
        Dependency(
            "org.example:offline-proxy",
            proxy_id,
            "proxies/A001_proxy.mov",
            required=False,
        ),
    )
    with production.transaction() as transaction:
        transaction.record_dependency_set(edit_id, dependencies)

    recorded = production.dependency_set(edit_id)  # None: never extracted
    if recorded is not None:
        print(f"{len(recorded.dependencies)} dependencies, {recorded.status.name}")
    return recorded


def all_dependencies(
    production: Production, edit_id: RepresentationId
) -> list[DependencyMatch]:
    matches: list[DependencyMatch] = []
    cursor = None
    while True:
        page = production.dependencies(
            edit_id, max_depth=1, max_representations=1000, limit=1, cursor=cursor
        )
        matches.extend(page.items)
        cursor = page.next_cursor
        if cursor is None:
            return matches


# [/dependency-set]


def main() -> None:
    if len(sys.argv) != 2:
        sys.exit("usage: provenance.py WORK_DIRECTORY")
    work = Path(sys.argv[1])
    media = work / "rushes" / "A001.mov"
    proxy = work / "proxies" / "A001_proxy.mov"
    proxy.parent.mkdir()
    proxy.write_bytes(b"proxy essence\n")
    edit = work / "edit" / "cut.otio"
    edit.parent.mkdir()
    edit.write_text("{}\n", encoding="utf-8")

    with Production.create(work / "provenance.pproj", "Provenance") as production:
        with production.transaction() as transaction:
            asset_id = transaction.import_media(media)
            edit_asset_id = transaction.import_media(edit)
        original = production.representations[asset_id][0]
        edit_representation = production.representations[edit_asset_id][0]
        with production.transaction() as transaction:
            proxy_id = transaction.add_single_file_representation(
                asset_id, RepresentationKind.PROXY, proxy
            )
        # Establish the caller's fingerprint domain before the activity, so
        # that the activity snapshots capture it.
        record_fingerprints(production, original, media)

        activity_id = record_proxy_generation(production, original.id, proxy_id)
        (activity,) = print_activities(production)
        assert activity.id == activity_id
        assert activity.tool == ToolIdentity("Example Transcoder", "3.2")
        assert activity.agent is not None
        assert activity.agent.identifier == ExternalIdentifier(
            "com.example.host", "node-4"
        )
        input_snapshot = activity.inputs[0].snapshot
        assert input_snapshot is not None
        assert {item.algorithm for item in input_snapshot.fingerprints} == {
            "example-sha256-representation",
            "pp-blake3-representation",
        }
        assert activity.outputs[0].snapshot is not None
        print_lineage(production, original.id)
        assert production.activities_consuming[original.id] == (activity,)
        assert production.provenance_descendants[original.id] == (proxy_id,)

        evaluation, _ = explain_proxy(production, proxy_id)
        assert evaluation.state is ArtifactKnowledgeState.CURRENT

        media.write_bytes(b"re-graded camera original\n")
        record_fingerprints(production, original, media)
        evaluation, reproducibility = explain_proxy(production, proxy_id)
        assert evaluation.state is ArtifactKnowledgeState.STALE
        assert [reason.kind for reason in evaluation.reasons] == [
            ArtifactReasonKind.FINGERPRINT_CHANGED
        ]
        assert evaluation.reasons[0].fingerprint_algorithm == (
            "example-sha256-representation"
        )
        assert not reproducibility.reproducible
        assert [issue.kind for issue in reproducibility.issues] == [
            ArtifactReproducibilityIssueKind.PARAMETERS_MISSING
        ]

        recorded = record_edit_dependencies(
            production, edit_representation.id, asset_id, original.id, proxy_id
        )
        assert recorded is not None
        assert recorded.status is DependencySetStatus.CURRENT
        assert len(recorded.dependencies) == 2
        assert [
            match.target
            for match in all_dependencies(production, edit_representation.id)
        ] == [
            asset_id,
            proxy_id,
        ]

        # A newer content observation of the edit invalidates its dependency set.
        edit.write_text('{"tracks": []}\n', encoding="utf-8")
        record_fingerprints(production, edit_representation, edit)
        updated = production.dependency_set(edit_representation.id)
        assert updated is not None
        assert updated.status is DependencySetStatus.NEEDS_EXTRACTION
        assert production.dependency_set(proxy_id) is None


if __name__ == "__main__":
    main()

File: docs/examples/cpp/provenance.cpp
// Runs the C++ listings about activities, artifact staleness, and dependencies.
//
// Each "[name]" ... "[/name]" region is included verbatim by the documentation
// build, so keep regions self-contained and readable. Usage:
//   postproject-cpp-provenance WORK_DIRECTORY
// The work directory is prepared by prepare-workdir.cmake.
#include <postproject/postproject.hpp>

#include <algorithm>
#include <cstdint>
#include <filesystem>
#include <fstream>
#include <iostream>
#include <optional>
#include <stdexcept>
#include <string>
#include <vector>

namespace {

void require(bool condition, const char *message) {
  if (!condition) {
    throw std::runtime_error(message);
  }
}

void write_file(const std::filesystem::path &path, const std::string &bytes) {
  std::filesystem::create_directories(path.parent_path());
  std::ofstream stream(path, std::ios::binary | std::ios::trunc);
  stream << bytes;
  require(static_cast<bool>(stream), "write file");
}

void print_snapshot(
    const char *label,
    const std::optional<postproject::ActivityEdgeSnapshot> &snapshot) {
  if (!snapshot) {
    std::cout << "  " << label << ": no snapshot\n";
    return;
  }
  for (const auto &fingerprint : snapshot->fingerprints) {
    std::cout << "  " << label << " at revision " << snapshot->revision_sequence
              << ": " << fingerprint.algorithm << " v" << fingerprint.version
              << '\n';
  }
}

// [activity-snapshots]
postproject::Uuid record_transcode(postproject::Production &production,
                                   const postproject::Uuid &original_id,
                                   const postproject::Uuid &proxy_id) {
  postproject::ActivitySpec spec{};
  spec.kind = "org.postproject:transcode";
  spec.started_at_unix_micros = 1'700'000'000'000'000;
  spec.finished_at_unix_micros = 1'700'000'060'000'000;
  spec.inputs = {{original_id, "org.postproject:essence"}};
  spec.outputs = {{proxy_id, std::nullopt}};
  spec.tool = postproject::ToolIdentity{"Example Encoder", "5.1", std::nullopt};
  spec.agent = postproject::AgentIdentity{
      "render-node-04", postproject::ExternalIdentifier{"com.example.farm.node",
                                                        "04", std::nullopt}};

  auto transaction = production.beginTransaction();
  const auto activity_id = transaction.createActivity(spec);
  transaction.commit();

  // Every edge carries the fingerprints its representation had at creation.
  for (const auto &activity : production.activities()) {
    std::cout << activity.kind << " by "
              << (activity.tool ? activity.tool->name : "unknown tool")
              << " on "
              << (activity.agent ? activity.agent->name.value_or("?") : "?")
              << '\n';
    for (const auto &input : activity.inputs) {
      print_snapshot("input", input.snapshot);
    }
    for (const auto &output : activity.outputs) {
      print_snapshot("output", output.snapshot);
    }
  }

  const auto consumers = production.activitiesConsuming(original_id);
  std::cout << "activities reading the original: " << consumers.size() << '\n';
  // Whole-set traversal, then the bounded paged form of the same query.
  const auto derived = production.descendants(original_id);
  const auto page = production.descendants(original_id, 8, 1000, 100);
  std::cout << derived.size() << " descendant(s), first page "
            << page.items.size() << '\n';
  return activity_id;
}
// [/activity-snapshots]

// [stale-after-change]
postproject::ArtifactEvaluation evaluate_after_change(
    postproject::Production &production,
    const postproject::Representation &original,
    const postproject::Fingerprint &new_resource_fingerprint,
    const postproject::Fingerprint &new_representation_fingerprint,
    const postproject::Uuid &proxy_id) {
  // Record what the host observed after the original was re-exported.
  auto transaction = production.beginTransaction();
  transaction.recordResourceFingerprint(original.resources.front().id,
                                        new_resource_fingerprint);
  transaction.recordRepresentationFingerprint(original.id,
                                              new_representation_fingerprint);
  transaction.commit();

  const auto evaluation = production.evaluateArtifact(proxy_id);
  if (evaluation.state == postproject::ArtifactKnowledgeState::stale) {
    for (const auto &reason : evaluation.reasons) {
      std::cout << "stale because " << static_cast<std::uint32_t>(reason.kind)
                << '\n';
    }
  }

  // Reproducibility says whether the recorded activity suffices to redo it.
  const auto reproducibility = production.artifactReproducibility(proxy_id);
  for (const auto &issue : reproducibility.issues) {
    std::cout << "cannot reproduce: " << static_cast<std::uint32_t>(issue.kind)
              << '\n';
  }
  return evaluation;
}
// [/stale-after-change]

// [dependency-set]
std::vector<postproject::DependencyMatch> record_scene_dependencies(
    postproject::Production &production, const postproject::Uuid &scene_id,
    const std::vector<postproject::Dependency> &observed) {
  // A dependency set replaces the complete previous observation.
  auto transaction = production.beginTransaction();
  transaction.recordDependencySet(scene_id, observed);
  transaction.commit();

  if (const auto set = production.dependencySet(scene_id)) {
    std::cout << set->dependencies.size() << " dependencies, "
              << (set->status == postproject::DependencySetStatus::current
                      ? "current"
                      : "needs extraction")
              << '\n';
  }

  std::vector<postproject::DependencyMatch> matches;
  std::optional<std::string> cursor;
  do {
    const auto page = production.dependencies(scene_id, 4, 1000, 1, cursor);
    matches.insert(matches.end(), page.items.begin(), page.items.end());
    cursor = page.next_cursor;
  } while (cursor.has_value());
  return matches;
}
// [/dependency-set]

postproject::Representation find(const postproject::Production &production,
                                 const postproject::Uuid &asset_id,
                                 const postproject::Uuid &id) {
  for (auto &representation : production.representations(asset_id)) {
    if (representation.id == id) {
      return representation;
    }
  }
  throw std::runtime_error("representation not found");
}

// Stand-in for the digest the host's hasher computes for changed bytes.
postproject::Fingerprint changed(const postproject::Fingerprint &stored,
                                 std::uint8_t fill) {
  return {stored.algorithm, stored.version,
          std::vector<std::uint8_t>(stored.value.size(), fill)};
}

} // namespace

int main(int argc, char **argv) {
  if (argc != 2) {
    std::cerr << "usage: postproject-cpp-provenance WORK_DIRECTORY\n";
    return 2;
  }
  const std::filesystem::path work = argv[1];
  const std::string media = (work / "rushes" / "A001.mov").string();

  try {
    write_file(work / "proxies" / "A001_proxy.mov", "proxy bytes");
    write_file(work / "characters" / "lead.usd", "#usda 1.0");
    write_file(work / "scenes" / "harbour.usd", "#usda 1.0 scene");

    auto production = postproject::Production::create(
        (work / "provenance.pproj").string(), "Provenance");
    auto setup = production.beginTransaction();
    const auto asset_id = setup.importMedia(media, "Camera A");
    const auto lead_asset_id =
        setup.importMedia((work / "characters" / "lead.usd").string());
    const auto scene_asset_id =
        setup.importMedia((work / "scenes" / "harbour.usd").string());
    setup.commit();
    const auto original_id = production.representations(asset_id).front().id;
    const auto lead_id = production.representations(lead_asset_id).front().id;
    const auto scene_id = production.representations(scene_asset_id).front().id;
    auto proxy_setup = production.beginTransaction();
    const auto proxy_id = proxy_setup.addSingleFileRepresentation(
        asset_id, postproject::RepresentationKind::proxy,
        (work / "proxies" / "A001_proxy.mov").string());
    proxy_setup.commit();

    const auto activity_id =
        record_transcode(production, original_id, proxy_id);
    const auto activities = production.activities();
    require(activities.size() == 1 && activities.front().id == activity_id,
            "one activity");
    const auto &activity = activities.front();
    require(activity.tool &&
                activity.tool->version == std::optional<std::string>("5.1"),
            "tool identity");
    require(activity.agent && activity.agent->identifier &&
                activity.agent->identifier->value == "04",
            "agent identity");
    require(activity.inputs.size() == 1 && activity.inputs.front().snapshot &&
                !activity.inputs.front().snapshot->fingerprints.empty(),
            "input snapshot");
    require(activity.outputs.size() == 1 && activity.outputs.front().snapshot,
            "output snapshot");
    require(production.activitiesConsuming(original_id).size() == 1,
            "consuming activity");
    require(production.descendants(original_id) == std::vector{proxy_id},
            "whole-set descendants");
    const auto page = production.descendants(original_id, 8, 1000, 100);
    require(page.items.size() == 1 &&
                page.items.front().object.id == proxy_id &&
                page.items.front().depth == 1,
            "paged descendants");
    require(production.evaluateArtifact(proxy_id).state ==
                postproject::ArtifactKnowledgeState::current,
            "proxy current before the change");

    write_file(media, "re-exported camera original");
    const auto original = find(production, asset_id, original_id);
    const auto evaluation = evaluate_after_change(
        production, original,
        changed(original.resources.front().fingerprints.front(), 0x55),
        changed(original.fingerprints.front(), 0x66), proxy_id);
    require(evaluation.state == postproject::ArtifactKnowledgeState::stale,
            "proxy stale");
    require(std::any_of(
                evaluation.reasons.begin(), evaluation.reasons.end(),
                [](const postproject::ArtifactReason &reason) {
                  return reason.kind ==
                         postproject::ArtifactReasonKind::fingerprint_changed;
                }),
            "fingerprint changed reason");
    const auto reproducibility = production.artifactReproducibility(proxy_id);
    require(!reproducibility.reproducible &&
                std::any_of(
                    reproducibility.issues.begin(),
                    reproducibility.issues.end(),
                    [](const postproject::ArtifactReproducibilityIssue &issue) {
                      return issue.kind ==
                             postproject::ArtifactReproducibilityIssueKind::
                                 parameters_missing;
                    }),
            "missing parameters");

    const std::vector<postproject::Dependency> observed{
        {std::nullopt,
         "org.example:plate",
         {postproject::ObjectKind::asset, asset_id},
         original_id,
         true,
         "rushes/A001.mov"},
        {std::nullopt,
         "org.example:character-reference",
         {postproject::ObjectKind::asset, lead_asset_id},
         lead_id,
         false,
         "characters/lead.usd"},
    };
    const auto matches =
        record_scene_dependencies(production, scene_id, observed);
    require(matches.size() == 2, "two dependencies paged");
    const auto current = production.dependencySet(scene_id);
    require(current &&
                current->status == postproject::DependencySetStatus::current &&
                current->dependencies.size() == 2 &&
                current->dependencies[1].authored_reference ==
                    "characters/lead.usd",
            "current dependency set");

    auto observation = production.beginTransaction();
    const auto scene = find(production, scene_asset_id, scene_id);
    observation.recordRepresentationFingerprint(
        scene_id, changed(scene.fingerprints.front(), 0x77));
    observation.commit();
    const auto after = production.dependencySet(scene_id);
    require(after && after->status ==
                         postproject::DependencySetStatus::needs_extraction,
            "dependency set needs extraction");
  } catch (const std::exception &error) {
    std::cerr << error.what() << '\n';
    return 1;
  }
}
